### ISSUE TITLE
Conversation or Dialog Indexing Example/Advice

### ISSUE THREAD

**Issue by AdolfVonKleist** [jina-ai/jina#3865]
**Describe your proposal/problem**
I'm interested in indexing spoken and transcribed, multi-party dialogs.  A key requirement that brought me to jina is the question of how to handle sequential queries that involve a 'he said and she replied' scenario.  For example:

  * Customer: Hello, I need some help with my product X
  * Agent: Hi I don't want to help you

I'm looking to implement a mechanism that identifies both sentiment/content of a particular _turn_ (easy), and then considers the _response_ from another participant in the conversation (unsolved).  A search concept might look something like:
  
  * `> customer:turn_type:greeting &  answer:agent:sentiment:negative`
  * `> customer:says:expletive & agent:answers:expletive`

these  are just some super trivial examples to convey the concept; we might like to track topic trends or other interaction series.  This sort of sequential filtering is not really supported in other engines like lucene or elastic or xapian, etc.; but I wonder if we could achieve something like this via jina?

**Comment by JoanFM** [comment]
Hello @AdolfVonKleist ,

Our `primitive` data types could support such type of data.

One thing I can imagine, could be considering a `Document` as a conversation and add to its `chunks` the ordered sequence of interaction:

Let's see a potential example.

```python
from jina import Document

doc = Document()
doc.chunks.append(text='customer text', tags={'interlocutor': 'customer'})
doc.chunks.append(text='agent response', tags={'interlocutor: 'agent'}) 
```

With this support u can have ur own custom embedding and indexing. 

I hope this can help, not sure what is the end to end solution you have in mind.

**Comment by AdolfVonKleist** [comment]
@JoanFM thanks for the reply!  Would the example you shared support the concept of adjacent or sequential seach?
We would want to be able to find document where `customer_text` is _followed by_ `agent_response`, but of course we don't know where in the sequence of turns this might occur, and it will be different for every document.  

One approach we have tried in the past with other frameworks is to cascade the retrieval process: 
 1. Find all documents that contain the target `customer_text` and `agent_response` (anywhere in the document) 
 2. Refine the results by checking that the `agent_response` actually follows the `customer_text` closely in the timeline.

however it would be nice to achieve both in a single, parsimonious query, and without otherwise unneeded duplication of index data.

In your example I am uncertain how I would indicate in the search that `customer_text` and `agent_response` should be _adjacent_ or close in the timeline to each other?

**Comment by JoanFM** [comment]
Hey @AdolfVonKleist ,

This is an interesting usecase.

Some things to clarify.

- Jina is not a closed solution, We at Jina try to provide you with the tools and design patterns that can help you build your tailored search solution.
- You can create your own Executors with your own logic to solve ur specific problem. All you need to do is learn `Flow`, `Executor` and `Document, DocumentArray` and start using them to build your solution. 

Let's try to have an example:

```python
import itertools
from jina import Flow, Executor, requests, Document, DocumentArray, DocumentArrayMemmap

class ConversationIndexer(Executor):

    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)
        self._docs = DocumentArrayMemmap(self.workspace)

    @requests(on='/index')
    def index(self, docs, **kwargs):
        # Index docs where the chunks are conversation snippets (Every conversation snippet is embedded through a transformer)
        self._docs.extend(docs)

    def _are_continuous(self, cust_doc, agent_doc, distance_between_sentences):
        return 0 < (agent_doc.tags['position'] - cust_doc.tags['position']) <= distance_between_sentences

    @requests(on='/search')
    def search(self, docs, parameters, **kwargs):
        # Very naive solution, if a better algorithm is found can easily be added here
        # Queries are docs where the first chunk is the customer_text and the second_chunk is the agent_response (All embedded through a Transformer)
        distance_between_sentences = parameters['distance_between_sentences']

        for query in docs:
            q_customer_text = DocumentArray(query.chunks[0], copy=True)
            q_agent_response = DocumentArray(query.chunks[1], copy=True)
            q_customer_text.match(self._docs, traversal_rpath=['c'], filter_fn=lambda c: c.tags['interlocutor'] == 'customer')
            q_agent_response.match(self._docs, traversal_rpath=['c'], filter_fn=lambda c: c.tags['interlocutor'] == 'agent')

            # now q_customer_text and q_agent_response have matching conversations, then u can look.
            for cust_match, agent_match in itertools.product(q_customer_text.matches, q_agent_response.matches):
                if self._are_continuous(cust_match, agent_match, distance_between_sentences):
                    query.matches.append(Document(chunks=[cust_match, agent_match]))

f = Flow().add(uses='jinahub://TransformerEncoder').add(uses=ConversationIndexer)
doc = Document()
doc.chunks.append(text='customer text', tags={'interlocutor': 'customer', 'position': 0})
doc.chunks.append(text='agent response', tags={'interlocutor': 'agent', 'position': 1})

query = Document()
query.chunks.append(text='customer text')
query.chunks.append(text='agent response')

with f:
    f.post(on='/index', inputs=[doc]) # index all ur conversation documents
    results = f.post(on='/search', inputs=[query], parameters={'distance_between_sentences': 1}, return_results=True)

print(f' {results[0].docs}')
```

This is just an example, and is not there to be reproduced, but this is what Jina allows, to encapsulate your own business logic into Executors with the help of all our design, abstractions and objects.

I hope this can help you to see the potential of Jina to find your own solution.

**Comment by AdolfVonKleist** [comment]
@JoanFM 

> Jina is not a closed solution, We at Jina try to provide you with the tools and design patterns that can help you build your tailored search solution.

Yes! This is exactly why I opened this question.  I am very amenable the idea of leveraging a reasonable pattern and doing even quite low-level customization myself.  But I was hoping for some pointers or a high-level assessment of whether my use case even made sense in the 'jina world', before diving down the rabbit hole.  You've provided exactly this.

One final followup:  in your example your 'document' has just 2 turns, and the search function also hardcodes the query texts to `0` and `1` indices in the `chunks` array.   Real conversations may have 100s of turns.  Would this already accommodate that scenario?  Or would I need to also iterate through the chunks array and compare all the sequential turn pairs?

**Comment by JoanFM** [comment]
> @JoanFM
> 
> > Jina is not a closed solution, We at Jina try to provide you with the tools and design patterns that can help you build your tailored search solution.
> 
> Yes! This is exactly why I opened this question. I am very amenable the idea of leveraging a reasonable pattern and doing even quite low-level customization myself. But I was hoping for some pointers or a high-level assessment of whether my use case even made sense in the 'jina world', before diving down the rabbit hole. You've provided exactly this.
> 
> One final followup: in your example your 'document' has just 2 turns, and the search function also hardcodes the query texts to `0` and `1` indices in the `chunks` array. Real conversations may have 100s of turns. Would this already accommodate that scenario? Or would I need to also iterate through the chunks array and compare all the sequential turn pairs?

What I built is just an example. I think you could see this as some sort of `meta-query-language` that u can customize urself using the `Document` structure and the `parameters` passed in the `Executor` methods.
